### ISSUE TITLE
Possible to exclude the testng-assert module when using testng

### ISSUE THREAD

**Issue by mpet** [testng-team/testng#3197]
### TestNG Version
v7.10.2

### Suggestion
Hi
In our team we discussed this:
Our majority of TestNG testcases uses https://joel-costigliola.github.io/assertj/ for assertions.
But when users write TestNG unit tests the first alternative that pops up is org.testng.Assert.
Would it be possible to make this module optional? This is especially important for users that
use TestNG for executing tests and assertj to assert. 

br,

//mikael

**Comment by juherr** [comment]
Currently, TestNG is released into a single artifact, but the project is already modularized and assertions are in their own module.
Maybe we should consider releasing TestNG artifact without assertions plus an extra artifact for assertions.

@krmahadevan WDYT

Related to #2649

**Comment by krmahadevan** [comment]
@juherr - That would mean that we would need to do 3 artifact releases

* Assertions library from TestNG
* TestNG without assertions
* TestNG with assertions (it would consume the assertion library)

Is this what we are looking at ?

**Comment by juherr** [comment]
Only assertions and testng without assertions.

It means the upgrade will need to add assertions as dependencies if used.

**Comment by krmahadevan** [comment]
Wouldn't that cause backward compatibility issues for users? Yes, we can send out announcements etc., but just calling out.

Should we do this as part of a minor release or a major release?

**Comment by martinaldrin** [comment]
I agree, release the assert library as a separate artifact. 
Mark it as optional dependency in the pom.xml that user can add if they need it.

**Comment by juherr** [comment]
@krmahadevan yes, it will make a little regression that must be documented

**Comment by mpet** [comment]
@juherr is it accepted?

**Comment by juherr** [comment]
@mpet I think it will accepted when someone will propose a pull request for that.
😉

**Comment by mpet** [comment]
@juherr  I guess we need to exclude the asserts from shaded jar too. Right?

**Comment by juherr** [comment]
@mpet excluded from the shaded jar and deployed as a standalone artifact.

**Comment by mpet** [comment]
ok. Any input on what should be in the build file for testng-asserts. I guess some of the stuff in 
https://github.com/testng-team/testng/blob/master/build.gradle.kts should be reused.

**Comment by mpet** [comment]
I cannot see where the 'publish' is done

**Comment by mpet** [comment]
@krmahadevan do you know the release process and which parts that I need?

**Comment by krmahadevan** [comment]
@mpet - The release process is documented here: https://github.com/testng-team/testng/wiki/TestNG-release-process

I believe that we would perhaps need to create a github actions workflow that would release assert library.

@vlsi - It would be good if you could help share some guidance around here.

**Comment by mpet** [comment]
yes it seems to have a dedicated flow where you set the RC  to be released. Not sure what is need in the build to support this. But I hope @vlsi can shed some light in this matter...

**Comment by mpet** [comment]
@krmahadevan is @vlsi still active? How can we make this go forward...

**Comment by krmahadevan** [comment]
@mpet - Vlad helped setup the build system for TestNG. He can help more aptly. Given that this is a holiday season I would expect latency in people responding. 

I am not well versed with Gradle as a build system so I don't have an answer and will need to spend time trying to figure out. 

So we wait till @vlsi gets back or you can try taking a jab at this.

**Comment by juherr** [comment]
Thanks for the feedback.
I think it could go faster 

What if "org.testng.assertions.Assert" is introduced and "org.testng.Assert" is deprecated in the next minor release?

What if "org.testng.assertions" is moved into its dedicated artifact, and core depends on it in the next minor release?

Both moves should not break anything, right?

**Comment by vlsi** [comment]
> What if "org.testng.assertions.Assert" is introduced and "org.testng.Assert" is deprecated in the next minor release?

Sure it is possible to combine the changes. I just assumed it would take some time to add `org.testng.assertions`, etc, etc.

---

By the way, my plan missed a release that would exclude "testng-assertions" from the default dependency of testng.jar.
Frankly, I am not sure if moving `org.testng.Assert` to its own jar is backward-compatible as the users might have used something like `-classpath testng.jar`

**Comment by juherr** [comment]
Right. It can break with cli runner.

Then I propose to create the new artifact with the new package, and deprecate assertions from the base package in a next minor version.

And remove deprecated in the next major as usual.

Wdyt @krmahadevan ?

**Comment by vlsi** [comment]
> In TestNG 7.13.0 we remove org.testng.Assert from the testng uber jar and be done with this activity

I would ask you not doing so.
It takes time for the consumers to adapt, so "removal" should better be delayed by 5 or so years.

**Comment by krmahadevan** [comment]
@vlsi Personally I wouldn't bother removing Asserts at all. A test runner is supposed to provide the basic assertion capabilities. If someone doesn't want to use them then they can always use something else. It's just a matter of preference. 

But the ask in this issue itself was that, they felt it was not an optimal experience. If we are going to keep the assertion classes for that long then we might as well not remove them 🫣

I don't have a strong opinion on this either way because there are straightforward ways of avoiding it, ranging from code reviews that enforce this avoidance to using the enforcer plug-in to ensure that the testng powered asserts are not used in the code base. 

I was merely trying to figure out the list of activities that need to be done ( along with some timelines ) to accomplish the ask in this issue.

**Comment by mpet** [comment]
@vlsi we have a large code base with many maven modules where we use testng as the testing  fwk. However there is a mix of asserts libraries used. Even if we recommend using assertj various IDE:s propose different asserts. People also copy paste code.

So we basically want to clean code.  Module by module and use one type of assert, assertj.
So what @juherr suggested seems very reasonable and  would support  us on our work to remove testng asserts. It would also give complie time error when users use testng asserts in the future.
So this change would be greatly appreciated.

**Comment by ben-manes** [comment]
I'd recommend instead maintaining this company policy by including common IDE configurations through build or repository, so that it is applied at project import. See Eclipse's _Type Filters_ and IntelliJ's _Auto-Import Excludes_. Then use an lint rule (an enforcer plugin, checkstyle, etc) to ban the import during the build so that it doesn't sneak through.
